pkg/supervisor/supervisor.go

```diff
@@ -0,0 +1,247 @@
+	s.shuttingDown = true
+	s.changed.Broadcast()
+}
+
+func (s *Supervisor) dependents(worker *Worker) (result []*Worker) {
+	for _, n := range s.names {
+		w := s.workers[n]
+		for _, r := range w.Requires {
+			if r == worker.Name {
+				result = append(result, w)
+				break
+			}
+		}
+	}
+	return
+}
+
+// make sure anything that would like to be running is actually
+// running
+func (s *Supervisor) reconcile() {
+	if s.shuttingDown {
+		s.reconcileShutdown()
+	} else {
+		s.reconcileNormal()
+	}
+}
+
+// shutdown anything that is safe to shutdown and not already shutdown
+func (s *Supervisor) reconcileShutdown() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process != nil && !w.process.shutdownClosed {
+			for _, d := range s.dependents(w) {
+				if s.workers[d.Name].process != nil {
+					log.Printf("cannot shutdown %s, %s still running", n, d.Name)
+					continue OUTER
+				}
+			}
+			close(w.process.shutdown)
+			w.process.shutdownClosed = true
+		}
+	}
+}
+
+// launch anything that is safe to launch and not already launched
+func (s *Supervisor) reconcileNormal() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process == nil {
+			for _, r := range w.Requires {
+				process := s.workers[r].process
+				if process == nil || !process.ready {
+					log.Printf("cannot start %s, %s not ready", n, r)
+					continue OUTER
+				}
+			}
+			s.launch(w)
+		}
+	}
+}
+
+func (s *Supervisor) launch(worker *Worker) {
+	context, cancel := context.WithCancel(s.context)
+	process := &Process{
+		Supervisor: s,
+		Worker:     worker,
+		Context:    context,
+		cancel:     cancel,
+		shutdown:   make(chan struct{}),
+	}
+	worker.process = process
+	go func() {
+		var err error
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					err = errors.Errorf("PANIC: %v", r)
+				}
+			}()
+			err = worker.Work(process)
+		}()
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+		worker.process = nil
+		if err != nil {
+			process.Log(err)
+			if worker.Retry {
+				if s.shuttingDown {
+					s.remove(worker)
+				} else {
+					process.Log("retrying...")
+				}
+			} else {
+				s.remove(worker)
+				s.errors = append(s.errors, err)
+				s.shuttingDown = true
+			}
+		} else {
+			s.remove(worker)
+		}
+		s.changed.Broadcast()
+	}()
+}
+
+type Process struct {
+	Supervisor *Supervisor
+	Worker     *Worker
+	// Used for hard cancel.
+	Context context.Context
+	cancel  context.CancelFunc
+	// Used to signal graceful shutdown.
+	shutdown       chan struct{}
+	ready          bool
+	shutdownClosed bool
+}
+
+// Invoked by a worker to signal it is ready.
+func (p *Process) Ready() {
+	p.Supervisor.mutex.Lock()
+	defer p.Supervisor.mutex.Unlock()
+	p.ready = true
+	p.Supervisor.changed.Broadcast()
+}
+
+// Used for graceful shutdown...
+func (p *Process) Shutdown() <-chan struct{} {
+	return p.shutdown
+}
+
+// Used for logging...
+func (p *Process) Log(obj interface{}) {
+	log.Printf("%s: %v", p.Worker.Name, obj)
```

Comment:
I don't like that this hard-codes the global `log.std` logger. IMO, that would be a blocker to using it in APro.  I would give the supervisor an optional Logger setting that takes an `interface { Printf(format string, v ...interface{}) }`; or maybe even a LoggerFactory that is something like

```go
type Logger interface {
    Printf(format string, v ...interface{})
}
type LoggerFactory interface {
    NewLogger(workerName string) Logger
}
```